File: python/monad/symbolic_engine.py
# Manipulation Monad Symbolic Engine Functions (Python version)
import tensorflow as tf
import math
import numpy as np
import logging

logger = logging.getLogger(__name__)

# --- Constants ---
OPERATION_COSTS = {
    'rotateBits': 0.05,
    'foldStructure': 0.08,
    'primeProbe': 0.12,
    'xorReflect': 0.07,
    'shiftRadix': 0.04,
    'noop': 0.0,
    'identity': 0.0,
    'customOpX': 0.10,
    'default': 0.01
}

ATTRACTORS = [
    "110110.110110", # 108-like
    "100100.100100", # 144-like
    "110110110.110", # 432-like
    "111000.000111", # mirror states
    "101010.010101", # binary harmonics
]

# --- Functions ---

def derive_sigma(tensor, radix_position='middle'):
    """
    Derives the symbolic state string sigma from a model tensor.
    Warning: Uses .numpy(), which can impact performance if called frequently within a graph.

    Args:
        tensor (tf.Tensor): Input tensor, assumed to be [batch?, ..., features].
                            Uses first batch item if batched.
        radix_position (str): Where to place the radix ('middle', etc.). Defaults to 'middle'.

    Returns:
        str | None: Sigma string "s_L.s_R" or None on error.
    """
    try:
        if not isinstance(tensor, tf.Tensor):
            logger.error("symbolic_engine.derive_sigma: Input is not a TensorFlow tensor.")
            return None
        if tf.size(tensor) == 0:
            logger.warning("symbolic_engine.derive_sigma: Input tensor is empty.")
            return "." # Return empty parts

        # Handle potential batch dimension
        tensor_to_process = tensor
        if len(tensor.shape) > 1 and tensor.shape[0] > 1:
            tensor_to_process = tensor[0:1, ...] # Take first item

        # Flatten and get numpy array - SYNCHRONOUS
        flat_values = tf.reshape(tensor_to_process, [-1]).numpy()

        if flat_values.size == 0:
            return "."

        min_val = np.min(flat_values)
        max_val = np.max(flat_values)
        val_range = max_val - min_val

        # Binarize
        if val_range == 0:
            binary_bits = ['1' if v >= 0.5 else '0' for v in flat_values]
        else:
            binary_bits = ['1' if (v - min_val) / val_range >= 0.5 else '0' for v in flat_values]

        # Place radix point
        if radix_position == 'middle':
            mid_point = len(binary_bits) // 2
        else: # Default to middle
            mid_point = len(binary_bits) // 2

        s_l = "".join(binary_bits[:mid_point])
        s_r = "".join(binary_bits[mid_point:])
        return f"{s_l}.{s_r}"

    except Exception as e:
        logger.error(f"symbolic_engine.derive_sigma Error: {e}", exc_info=True)
        return None

def calculate_shannon_entropy(binary_string):
    """Calculates Shannon entropy H(s) for a binary string."""
    if not binary_string: return 0.0
    length = len(binary_string)
    if length == 0: return 0.0

    ones = binary_string.count('1')
    p1 = ones / length
    p0 = 1.0 - p1

    entropy = 0.0
    if p1 > 0: entropy -= p1 * math.log2(p1)
    if p0 > 0: entropy -= p0 * math.log2(p0)

    return entropy if not math.isnan(entropy) else 0.0

def compute_entropy(sigma, transformation_history=None):
    """
    Calculates symbolic entropy E(sigma) = H(sigma) + operation costs.

    Args:
        sigma (str): The symbolic state string "s_L.s_R".
        transformation_history (list, optional): List of dicts {'op': name, 'params': ...}. Defaults to [].

    Returns:
        float: The symbolic entropy E(sigma).
    """
    if transformation_history is None:
        transformation_history = []
    if sigma is None: return 0.0 # Handle null sigma

    parts = sigma.split('.')
    s_l = parts[0] if len(parts) > 0 else ""
    s_r = parts[1] if len(parts) > 1 else ""

    # Calculate Shannon entropy (average of left/right parts)
    h_l = calculate_shannon_entropy(s_l)
    h_r = calculate_shannon_entropy(s_r)
    base_entropy = (h_l + h_r) / 2.0

    # Add costs of transformations
    operation_cost_sum = sum(
        OPERATION_COSTS.get(t.get('op', 'default'), OPERATION_COSTS['default'])
        for t in transformation_history
    )

    return base_entropy + operation_cost_sum

def are_complementary(str1, str2):
    """Checks if two binary strings are bitwise complements."""
    if not str1 or not str2 or len(str1) != len(str2):
        return False
    return all(c1 != c2 for c1, c2 in zip(str1, str2))

def calculate_phase_match(sigma, p):
    """Calculates phase match score phi_p(sigma)."""
    if sigma is None: return 0.0
    binary_string = sigma.replace('.', '')
    n = len(binary_string)
    if p <= 0 or n < 2 * p:
        return 0.0

    total_score = 0.0
    window_count = 0

    for i in range(n - 2 * p + 1):
        pattern1 = binary_string[i : i + p]
        pattern2 = binary_string[i + p : i + 2 * p]
        if pattern1 == pattern2:
            total_score += 1.0
        elif are_complementary(pattern1, pattern2):
            total_score += 0.5
        window_count += 1

    return total_score / window_count if window_count > 0 else 0.0

def calculate_resonance(sigma, primes):
    """
    Calculates Resonance coefficient R(sigma).

    Args:
        sigma (str): The symbolic state string "s_L.s_R".
        primes (list): Array of primes to check resonance against.

    Returns:
        float: The resonance coefficient R(sigma).
    """
    if sigma is None: return 0.0
    resonance_sum = 0.0
    for p in primes:
        if p > 0:
            phi_p = calculate_phase_match(sigma, p)
            resonance_sum += (1.0 / p) * phi_p
    return resonance_sum

def compute_parity(sigma):
    """Calculates Parity(sigma)."""
    if sigma is None: return None
    ones = sigma.replace('.', '').count('1')
    return "even" if ones % 2 == 0 else "odd"

def hamming_distance(s1, s2):
    """Calculates Hamming distance between two binary strings (padded)."""
    if s1 is None or s2 is None: return float('inf') # Or handle differently
    b1 = s1.replace('.', '')
    b2 = s2.replace('.', '')
    len1, len2 = len(b1), len(b2)
    max_len = max(len1, len2)
    p1 = b1.ljust(max_len, '0')
    p2 = b2.ljust(max_len, '0')

    hamming = sum(c1 != c2 for c1, c2 in zip(p1, p2))
    return hamming

def symbolic_distance(sigma1, sigma2, entropy_weight=0.1):
    """
    Calculates entropy-weighted symbolic distance D(sigma1, sigma2).

    Args:
        sigma1 (str): First sigma string.
        sigma2 (str): Second sigma string.
        entropy_weight (float, optional): Weight for the entropy penalty. Defaults to 0.1.

    Returns:
        float: The symbolic distance.
    """
    if sigma1 is None or sigma2 is None: return float('inf') # Or handle differently
    ham_dist = hamming_distance(sigma1, sigma2)

    # Calculate Shannon entropy on the full binary string representation
    e1 = calculate_shannon_entropy(sigma1.replace('.', ''))
    e2 = calculate_shannon_entropy(sigma2.replace('.', ''))
    entropy_penalty = abs(e1 - e2)

    return ham_dist + entropy_weight * entropy_penalty

def get_collapse_value(sigma, attractor_list, entropy_weight):
    """
    Finds the nearest attractor string to sigma based on symbolic distance.

    Args:
        sigma (str): The current sigma string.
        attractor_list (list): The list of attractor strings.
        entropy_weight (float): Weight for entropy penalty in distance.

    Returns:
        str | None: The nearest attractor string or None if list is empty or sigma is None.
    """
    if sigma is None or not attractor_list:
        return None

    min_d = float('inf')
    best_attractor = None
    for attractor in attractor_list:
        d = symbolic_distance(sigma, attractor, entropy_weight)
        if d < min_d:
            min_d = d
            best_attractor = attractor
    return best_attractor

def is_collapsed(sigma, entropy, resonance, parity, config):
    """
    Checks if a sigma state meets the collapse conditions.

    Args:
        sigma (str): The sigma string.
        entropy (float): Pre-calculated E(sigma).
        resonance (float): Pre-calculated R(sigma).
        parity (str): Pre-calculated Parity(sigma).
        config (dict): Configuration object with thresholds
                       ('entropy_min', 'entropy_max', 'resonance_threshold').

    Returns:
        bool: True if collapsed, False otherwise.
    """
    if sigma is None: return False
    meets_resonance = resonance >= config.get('resonance_threshold', 0.7)
    meets_entropy = (config.get('entropy_min', 0.2) <= entropy <= config.get('entropy_max', 0.3))
    meets_parity = parity == 'even'
    return meets_resonance and meets_entropy and meets_parity

# Export dict (less common in Python, users import specific functions)
# __all__ = [
#     'derive_sigma', 'compute_entropy', 'calculate_shannon_entropy',
#     'compute_parity', 'calculate_resonance', 'calculate_phase_match',
#     'are_complementary', 'symbolic_distance', 'hamming_distance',
#     'get_collapse_value', 'is_collapsed', 'OPERATION_COSTS', 'ATTRACTORS'
# ]

File: js/monad/symbolicEngine.js
// Manipulation Monad Symbolic Engine Functions
// Based on provided formalism and definitions.

const tf = require('@tensorflow/tfjs-node'); // Keep for deriveSigma

const operationCosts = {
  rotateBits: 0.05,
  foldStructure: 0.08,
  primeProbe: 0.12,
  xorReflect: 0.07,
  shiftRadix: 0.04,
  noop: 0.0,
  identity: 0.0, // Added alias for clarity
  default: 0.01
};

const attractors = [
  "110110.110110", // 108-like
  "100100.100100", // 144-like
  "110110110.110", // 432-like
  "111000.000111", // mirror states
  "101010.010101", // binary harmonics
];

/**
 * Derives the symbolic state string sigma from a model tensor (e.g., hiddenStates).
 * @param {tf.Tensor} tensor - Input tensor, assumed to be [batch?, ..., features]. Uses first batch item if batched.
 * @returns {string | null} Sigma string "s_L.s_R" or null on error.
 */
function deriveSigma(tensor) {
  try {
    if (!tensor || typeof tensor.shape === 'undefined' || tensor.isDisposed) {
      console.error("symbolicEngine.deriveSigma: Invalid or disposed tensor input.");
      return null;
    }

    // Handle potential batch dimension - process first batch item for sigma
    let tensorToProcess = tensor;
    if (tensor.shape.length > 1 && tensor.shape[0] > 1) {
      tensorToProcess = tensor.slice(0, 1); // Take first item
    }

    const flatTensor = tensorToProcess.flatten();
    const values = flatTensor.dataSync(); // Synchronous!
    tf.dispose(flatTensor);
    if (tensorToProcess !== tensor) tf.dispose(tensorToProcess);

    if (values.length === 0) return "."; // Return empty parts

    const min = Math.min(...values);
    const max = Math.max(...values);
    const range = max - min;

    let binaryBits = [];
    if (range === 0) {
      binaryBits = values.map(v => (v >= 0.5 ? '1' : '0'));
    } else {
      binaryBits = values.map(v => ((v - min) / range >= 0.5 ? '1' : '0'));
    }

    // Place radix point at midpoint
    const midPoint = Math.floor(binaryBits.length / 2);
    const s_L = binaryBits.slice(0, midPoint).join('');
    const s_R = binaryBits.slice(midPoint).join('');
    return `${s_L}.${s_R}`;

  } catch (error) {
    console.error("symbolicEngine.deriveSigma Error:", error);
    if (tensor && !tensor.isDisposed) tf.dispose(tensor);
    return null;
  }
}

/** Calculates Shannon entropy H(s) for a binary string. */
function calculateShannonEntropy(binaryString) {
    if (!binaryString || binaryString.length === 0) return 0;
    const len = binaryString.length;
    let ones = 0;
    for (let i = 0; i < len; i++) {
        if (binaryString[i] === '1') ones++;
    }
    const p1 = ones / len;
    const p0 = 1 - p1;

    let entropy = 0;
    if (p1 > 0) entropy -= p1 * Math.log2(p1);
    if (p0 > 0) entropy -= p0 * Math.log2(p0);
    return isNaN(entropy) ? 0 : entropy;
}


/**
 * Calculates symbolic entropy E(sigma) = H(sigma) + operation costs.
 * @param {string} sigma - The symbolic state string "s_L.s_R".
 * @param {Array<object>} transformationHistory - Array of {op: name, params}.
 * @returns {number} The symbolic entropy E(sigma).
 */
function computeEntropy(sigma, transformationHistory = []) {
  const parts = sigma.split('.');
  const s_L = parts[0] || "";
  const s_R = parts[1] || "";

  // Calculate Shannon entropy (average of left/right parts)
  const h_L = calculateShannonEntropy(s_L);
  const h_R = calculateShannonEntropy(s_R);
  const baseEntropy = (h_L + h_R) / 2; // Using average as per previous implementation

  // Add costs of transformations
  let operationCostSum = 0;
  for (const transform of transformationHistory) {
    operationCostSum += (operationCosts[transform.op] || operationCosts.default);
  }

  return baseEntropy + operationCostSum;
}

/** Checks if two binary strings are bitwise complements. */
function areComplementary(str1, str2) {
  if (!str1 || !str2 || str1.length !== str2.length) return false;
  return str1.split('').every((v, i) => v !== str2[i]);
}

/** Calculates phase match score phi_p(sigma). */
function calculatePhaseMatch(sigma, p) {
    const binaryString = sigma.replace('.', '');
    if (p <= 0 || binaryString.length < 2 * p) return 0;

    let totalScore = 0;
    let windowCount = 0;

    for (let i = 0; i <= binaryString.length - 2 * p; i++) {
        const pattern1 = binaryString.substring(i, i + p);
        const pattern2 = binaryString.substring(i + p, i + 2 * p);
        if (pattern1 === pattern2) {
            totalScore += 1.0;
        } else if (areComplementary(pattern1, pattern2)) {
            totalScore += 0.5;
        }
        windowCount++;
    }
    return windowCount > 0 ? totalScore / windowCount : 0;
}

/**
 * Calculates Resonance coefficient R(sigma).
 * @param {string} sigma - The symbolic state string "s_L.s_R".
 * @param {Array<number>} primes - Array of primes to check resonance against.
 * @returns {number} The resonance coefficient R(sigma).
 */
function calculateResonance(sigma, primes) {
  let resonanceSum = 0;
  for (const p of primes) {
    if (p > 0) {
      const phi_p = calculatePhaseMatch(sigma, p);
      resonanceSum += (1 / p) * phi_p;
    }
  }
  return resonanceSum;
}

/** Calculates Parity(sigma). */
function computeParity(sigma) {
  const ones = (sigma.match(/1/g) || []).length;
  return ones % 2 === 0 ? "even" : "odd";
}

/** Calculates Hamming distance between two binary strings (padded). */
function hammingDistance(s1, s2) {
    const b1 = s1.replace('.', '');
    const b2 = s2.replace('.', '');
    const maxLen = Math.max(b1.length, b2.length);
    const p1 = b1.padEnd(maxLen, '0');
    const p2 = b2.padEnd(maxLen, '0');

    let hamming = 0;
    for (let i = 0; i < maxLen; i++) {
        if (p1[i] !== p2[i]) hamming++;
    }
    return hamming;
}

/**
 * Calculates entropy-weighted symbolic distance D(sigma1, sigma2).
 * @param {string} sigma1 - First sigma string.
 * @param {string} sigma2 - Second sigma string.
 * @param {number} [entropyWeight=0.1] - Weight for the entropy penalty.
 * @returns {number} The symbolic distance.
 */
function symbolicDistance(sigma1, sigma2, entropyWeight = 0.1) {
  const hamming = hammingDistance(sigma1, sigma2);

  // Calculate Shannon entropy on the full binary string representation
  const e1 = calculateShannonEntropy(sigma1.replace('.', ''));
  const e2 = calculateShannonEntropy(sigma2.replace('.', ''));
  const entropyPenalty = Math.abs(e1 - e2);

  return hamming + entropyWeight * entropyPenalty;
}

/**
 * Finds the nearest attractor string to sigma based on symbolic distance.
 * @param {string} sigma - The current sigma string.
 * @param {Array<string>} attractorList - The list of attractor strings.
 * @param {number} entropyWeight - Weight for entropy penalty in distance.
 * @returns {string | null} The nearest attractor string or null if list is empty.
 */
function getCollapseValue(sigma, attractorList, entropyWeight) {
  if (!attractorList || attractorList.length === 0) return null;

  let minD = Infinity;
  let bestAttractor = null;
  for (let attractor of attractorList) {
    const d = symbolicDistance(sigma, attractor, entropyWeight);
    if (d < minD) {
      minD = d;
      bestAttractor = attractor;
    }
  }
  return bestAttractor;
}

/**
 * Checks if a sigma state meets the collapse conditions.
 * @param {string} sigma - The sigma string.
 * @param {number} entropy - Pre-calculated E(sigma).
 * @param {number} resonance - Pre-calculated R(sigma).
 * @param {string} parity - Pre-calculated Parity(sigma).
 * @param {object} config - Configuration object with thresholds (entropyMin, entropyMax, resonanceThreshold).
 * @returns {boolean} True if collapsed, false otherwise.
 */
function isCollapsed(sigma, entropy, resonance, parity, config) {
  const meetsResonance = resonance >= config.resonanceThreshold;
  const meetsEntropy = entropy >= config.entropyMin && entropy <= config.entropyMax;
  const meetsParity = parity === 'even';
  return (sigma !== null && meetsResonance && meetsEntropy && meetsParity);
}

module.exports = {
  deriveSigma,
  computeEntropy,
  calculateShannonEntropy, // Export helper if needed elsewhere
  computeParity,
  calculateResonance,
  calculatePhaseMatch, // Export helper if needed
  areComplementary, // Export helper
  symbolicDistance,
  hammingDistance, // Export helper
  getCollapseValue,
  isCollapsed,
  operationCosts, // Export constants
  attractors      // Export constants
};